crates/host/src/secrets.rs

```diff
@@ -0,0 +1,188 @@
+//! Module with structs for use in managing and accessing secrets in a wasmCloud lattice
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use anyhow::ensure;
+use async_nats::{jetstream::kv::Store, Client};
+use futures::stream;
+use futures::stream::{StreamExt, TryStreamExt};
+use secrecy::Secret;
+use serde::{Deserialize, Serialize};
+use tokio::sync::RwLock;
+use tracing::instrument;
+use wasmcloud_runtime::capability::secrets::store::SecretValue;
+use wasmcloud_secrets_client::Client as WasmcloudSecretsClient;
+use wasmcloud_secrets_types::{Application, Context, Secret as WasmcloudSecret, SecretRequest};
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
+struct SecretReference {
+    /// The backend to use for retrieving the secret.
+    pub backend: String,
+    /// The key to use for retrieving the secret from the backend.
+    pub key: String,
+    /// The version of the secret to retrieve. If not supplied, the latest version will be used.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub version: Option<String>,
+}
+
+#[derive(Debug)]
+/// A manager for fetching secrets from a secret store, caching secrets clients for efficiency.
+pub struct Manager {
+    config_store: Store,
+    secret_store_topic: Option<String>,
+    nats_client: Client,
+    backend_clients: Arc<RwLock<HashMap<String, Arc<WasmcloudSecretsClient>>>>,
+}
+
+impl Manager {
+    /// Create a new secret manager with the given configuration store, secret store topic, and NATS client.
+    ///
+    /// All secret references will be fetched from this configuration store and the actual secrets will be
+    /// fetched by sending requests to the configured topic. If the provided secret_store_topic is None, this manager
+    /// will always return an error if [`Self::fetch_secrets`] is called with a list of secrets.
```

Comment:
Hmnn if the goal is making sure that the host always has a secrets manager it can call, can we solve that at the host level? It feels like `SecretsManager` shouldn't exist if the topic isn't there...

How do you feel about implementing trait stuff on `Option<SecretsManager>`? that way the host could have it but the ergonomics wouldn't be so bad?

Reply:
I don't mind that option, mild concern about ease of tracking down bugs with traits on an option, but I still do prefer the current method of a short-circuit just like we do for policy requests. I like keeping the pattern we have in place for now but let me know if you feel strongly about it